_posts/2024-04-20-websockets2.md

```diff
@@ -0,0 +1,1861 @@
+  def chatState: F[String]
+}
+```
+Similarly, we'll also need a companion object:
+
+```scala
+object ChatProtocol {
+
+  def make[F[_]](
+      redisP: RedisProtocol[F],
+      postgresP: PostgresProtocol[F]
+  ): ChatProtocol[F] = {
+    new ChatProtocol[F] {
+      ???
+    }
+  }
+}
+```
+
+### 6.1. The register() Function
+Let's start with the `register()` function in `ChatProtocol`. To register someone, we'll need to first check if the user name exists in Redis, then add it to both the Redis and Postgres databases:
+
+```scala
+new RedisProtocol[F] {
+  ...
+  override def usernameExists(username: String): F[Boolean] = {
+    redis.hVals("users").map { nameslist =>
+      nameslist.exists(u => u.toLowerCase == username.toLowerCase)
+    }
+  }
+}
+```
+The `usernameExists()` function takes a username of type `String` and returns an `F[Boolean]`. Here we use the `hVals()` function, which takes the name of the hash, and returns an `F[List[String]]` containing the list of values. We map on this value and call the `exists()` method which returns a `Boolean` depending on if a match is found or not. 
+
+Next we'll implement the `createUser()`, and `createRoom()` functions in `RedisProtocol[F]` since they are very similar:
+
+```scala
+import cats.syntax.all.*
+import cats.Monad
+
+object RedisProtocol {
+  def make[F[_]: Monad](
+      redis: RedisCommands[F, String, String]
+  ): RedisProtocol[F] = {
+    new RedisProtocol[F] {
+      override def createUser(user: User): F[Unit] = {
+        redis.hSet("users", user.toMap).void
+      }
+
+      override def createRoom(room: Room): F[Unit] = {
+        redis.hSet("rooms", room.toMap).void
+      }
+    }
+  }
+}
+```
+Here we use the `hSet()` function to create the `users` and `rooms` Redis hashes. The method takes a key value which we supplied as `users` and `rooms`, these are the names of the hashes, the second argument is the `field` which is of type `Map[String, String]`.
+
+Notice the new `toMap()` function on `Room` and `User`, that we used to create the `Map`. Let's implement them next:
+
+```scala
+  //in user.scala
+  case class User(id: UserId, name: UserName) {
+    def toMap = Map((id.id.toString, name.name))
+  }
+
+  //in room.scala
+  case class Room(id: RoomId, name: RoomName) {
+    def toMap = Map((id.id.toString, name.name))
+  }
+```
+The `toMap` method creates a `Map` with a tuple of the `UUID` string and the name from `UserName` and `RoomName` respectively. We update these methods in `user.scala` and `room.scala`.
+
+Now back the `ChatProtocol[F]` `register()` function:
+
+```scala
+import cats.Monad
+import cats.syntax.all.*
+import cats.effect.std.UUIDGen
+import cats.effect.kernel.Concurrent
+
+object ChatProtocol {
+  def make[F[_]: Monad: UUIDGen: Concurrent](
+      redisP: RedisProtocol[F],
+      postgresP: PostgresProtocol[F]
+  ): ChatProtocol[F] = {
+    new ChatProtocol[F] {
+      override def register(name: String): F[OutputMessage] = {
+        for
```

Comment:
let's not mix styles - please use curly braces
apply on all for-comprehensions and pattern matching

Reply:
Updated